### PR TITLE
Implement Exploratory Landscape Analysis (ELA) with Information Content (IC)

### PR DESCRIPTION
**Pull request by smitchaudhary**
Solves #646

### CODE REVIEW

**Comment by smitchaudhary** [comment]
@mlahariya Before the code review, do you think this belongs here and the directory structure is something you would suggest?

**Comment by mlahariya** [comment]
> @mlahariya Before the code review, do you think this belongs here and the directory structure is something you would suggest?

Seems good to me. Lets keep it here.

File: qadence/ml_tools/information/information_content.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+logger = getLogger("ml_tools")
+
+
+class InformationContent:
+    def __init__(
+        self,
+        model: nn.Module,
+        loss_fn: Callable,
+        xs: Any,
```
Comment (mlahariya) [inline qadence/ml_tools/information/information_content.py:21]: Why Any?

File: qadence/ml_tools/trainer.py [hunk 2]
```diff
@@ -711,3 +712,107 @@ def build_optimize_result(
...
+        """
+        if not self._use_grad:
+            logger.warning(
+                "Gradient norm bounds are only relevant when using a gradient based optimizer. \
+                    Currently the trainer is set to use a gradient-free optimizer."
+            )
+
+        dataloader = dataloader if dataloader is not None else self.train_dataloader
+
+        batch = next(iter(self._batch_iter(dataloader, num_batches=1)))
```
Comment (mlahariya) [inline qadence/ml_tools/trainer.py:809]: Why not use the Trainers _batch_iter to get a batch of data? It will take care of none/dataloaders as well as setting to device and everything. You can get a single batch of XS from the _batch_iter and break if you only need a single batch.

File: docs/tutorials/qml/ml_tools/trainer.md [hunk 3]
```diff
@@ -469,7 +469,85 @@ for i in range(n_epochs):
...
+    model=qnn,
+    optimizer=optimizer,
+    config=train_config,
+    loss_fn="mse",
+    train_dataloader=train_loader,
+    optimize_step=optimize_step,
+)
+
+# Perform exploratory landscape analysis with Information Content
+ic_sensitivirty_threshold = 1e-4
```
Comment (mlahariya) [inline docs/tutorials/qml/ml_tools/trainer.md:524]: ic_sensitivity_threshold

File: docs/tutorials/qml/ml_tools/trainer.md [hunk 4]
```diff
@@ -469,7 +469,85 @@ for i in range(n_epochs):
...
 
-### 6.4. Custom `train` loop
+### 6.4. Performing pre-training Exploratory Landscape Analysis (ELA) with Information Content (IC)
+
+Before one embarks on training a model, one may wish to analyze the loss landscape to judge the trainability and catch vanishing gradient issues early.
+One way of doing this is made possible via calculating the [Information Content of the loss landscape](https://www.nature.com/articles/s41534-024-00819-8).
+This is done by discretizing the gradient in the loss landscapes and then calculating the information content therein.
+This serves as a measure of flatness or ruggedness of the loss landscape.
+Quantitatively, the information content allows us to get bounds on the average norm of the gradient in the loss landscape.
+
```
Comment (mlahariya) [inline docs/tutorials/qml/ml_tools/trainer.md:479]: It would be nice if we can briefly explain here what the maximum IC and sensitivity IC bounds represent. 
(Or all 3 outputs of the calculate_grad_norm_bounds_ic function)

File: qadence/ml_tools/trainer.py [hunk 5]
```diff
@@ -711,3 +712,107 @@ def build_optimize_result(
         self.opt_result = OptimizeResult(
             self.current_epoch, self.model_old, self.optimizer_old, loss, metrics
         )
+
+    def calculate_grad_norm_bounds_ic(
```
Comment (mlahariya) [inline qadence/ml_tools/trainer.py:716]: Consider renaming to get_ic_bounds for simplicity.

File: qadence/ml_tools/information/information_content.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        gradient from the Information Content of the loss landscape.
+
+        Args:
+            model: The quantum or classical model to analyze.
+            loss_fn: Loss function that takes model output and calculates loss
+            xs: Input data to evaluate the model on
+            epsilons: The thresholds to use for discretization of the finite derivatives
+            variaton_multiple: The number of sets of variational parameters to generate per each
+                variational parameter. The number of variational parameters required for the
+                statisctiacal analysis scales linearly with the amount of them present in the
```
Comment (mlahariya) [inline qadence/ml_tools/information/information_content.py:38]: statistical

File: qadence/ml_tools/information/information_content.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+        This class handles the study of loss landscape from information theoric
+        perspective and provides methods to get bounds on the norm of the
+        gradient from the Information Content of the loss landscape.
+
+        Args:
+            model: The quantum or classical model to analyze.
+            loss_fn: Loss function that takes model output and calculates loss
+            xs: Input data to evaluate the model on
+            epsilons: The thresholds to use for discretization of the finite derivatives
+            variaton_multiple: The number of sets of variational parameters to generate per each
```
Comment (mlahariya) [inline qadence/ml_tools/information/information_content.py:36]: variation_multiple

**Review by mlahariya — COMMENTED** [review]
Hey! Overall it looks really good. Thanks for the great work. 

Just some minor comments.

File: qadence/ml_tools/information/information_content.py (hunk 1 above)
Comment (smitchaudhary) [reply qadence/ml_tools/information/information_content.py:21]: Depends on what the model and the loss functions want. For PINN problems, it could be a dictionary of tensors (coming from a dictdataloader), or just a Tensor if you have a Dataloader, or even None if the model does not take any inputs.

File: qadence/ml_tools/information/information_content.py (hunk 7 above)
Comment (smitchaudhary) [reply qadence/ml_tools/information/information_content.py]: Sorry about the embarrassing amount of typos. Fixed now.

File: qadence/ml_tools/trainer.py (hunk 2 above)
Comment (smitchaudhary) [reply qadence/ml_tools/trainer.py:809]: I am not sure what you mean here.

I wanted to allow the user to provide a dataloader for IC calculations. If they provided None, then the dataloader within the Trainer class is used (if that is None too, then it means the user intends a model which does not take any input).

And then, I use the _batch_iter to get the batch and so on.

File: docs/tutorials/qml/ml_tools/trainer.md [hunk 8]
```diff
@@ -469,7 +469,93 @@ for i in range(n_epochs):
...
+Using the information conetnt technique, we can get two types of bounds on the average of the norm of the gradient.
```
Comment (lorenzocardarelli) [inline docs/tutorials/qml/ml_tools/trainer.md:480]: conetnt --> content

File: docs/tutorials/qml/ml_tools/trainer.md [hunk 9]
```diff
@@ -469,7 +469,93 @@ for i in range(n_epochs):
...
+print(
+    f"Using sensitivity IC, the gradients are bounded above by {sensitivity_ic_upper_bound:.3f}"
+)
+
+# Resume training as usual...
+
+trainer.fit(train_loader)
+```
+
+The `get_ic_grad_bounds` functoin returns a tuple containing a tuple containing the lower bound as achieved in maximum IC case, upper bound as achieved in maximum IC case, and the upper bound for the sensitivity IC case.
```
Comment (lorenzocardarelli) [inline docs/tutorials/qml/ml_tools/trainer.md:552]: functoin --> function

**Comment by lorenzocardarelli** [comment]
On the algorithmic side: I checked some tensors shape and formulas, everything looks correct. Kudos for the documentation, it is well explained. A couple of typos to amend. Good to merge.

File: qadence/ml_tools/trainer.py (hunk 2 above)
Comment (mlahariya) [reply qadence/ml_tools/trainer.py:809]: Ah now i understand. This seems fair. 
I was wondering why do you check if the dataloader is None or not, but I think its fair the way you do it.
